01projects/sec04CMakeHelloWorld.md

```diff
@@ -0,0 +1,237 @@
+
+### Populating the top-level (root) `CMakeLists.txt`
+
+#### `cmake_minimum_required()`
+
+`CMakeLists.txt` files are plain-text files written in the CMake-specific custom language syntax.
+For every project, the top most `CMakeLists.txt` must start by specifying a minimum CMake version using the [`cmake_minimum_required()`](https://cmake.org/cmake/help/latest/command/cmake_minimum_required.html) command.
+Although upper, lower and mixed case commands are supported by CMake, the modern trend is to use lower case commands.
+Yet another thing to note is that, like C++ itself, `CMake` is whitespace-agnostic while processing `CMakeLists.txt` files.
+
+Therefore, we start off with the following:
+``` cmake
+cmake_minimum_required(VERSION 3.25)
+```
+
+Even the most recent versions of CMake installations can replicate the exact behaviour of older releases.
+In this example, the minimum `VERSION` is specified, and other options such as the maximum version the project is tested with have been omitted.
+
+Owing to its long development history and strong emphasis on retaining backward compatibility, most CMake commands have multiple variants and sometimes have an overwhelming array of options.
+Naturally, it is not feasible to cover every aspect of the language here.
+However, we shall skim just enough of the language here to address our needs, and developers typically acquire proficiency in the language gradually over the years.
+Therefore, we encourage you not to get disheartened by the slightly verbose documentation at the outset.
+
+The main caveat regarding CMake's version used here is that the chosen minimum version should at least support C++17 language standards and compiler options.
+Recent versions of CMake have improved package detection, simplified dependency handling.
+While an older version suffices for this simple HelloWorld example, it does not hurt to use a CMake version as recent as possible.
+
+#### `project()`
+
+Next, we use the `project()` command to set the project name.
+This is required with every project and should be called soon after `cmake_minimum_required()`.
+This command can also be used to specify other project level information such as the language or version number.
+
+``` cmake
+project(hello_world
+  VERSION 0.0.1
+  LANGUAGES CXX
+)
+```
+
+The first (required) argument to this command is the project's name.
+The name of the project can be any valid string.
+It does not have to be the same string as the name of the main executable program to be built (although it is conventional/convenient to use the same string for them).
+The VERSION field is optional, but is conventional to populate it.
+Next, we specify the programming languages that the project uses.
+The keyword for C++ projects is `CXX`.
+It is helpful to specify the list of languages used in the project, so that cmake does not spend time introspecting the system for the presence of the toolchains for every supported language (CUDA, fortran, C, Java etc).
```

Comment:
```suggestion
It is helpful to specify the list of languages used in the project, so that cmake does not spend time inspecting the system for the presence of the toolchains for every supported language (CUDA, fortran, C, Java etc).
```

Reply:
@mmcleod89 It is actually called system introspection in CMake terminology (official online docs as well as in multiple textbooks that I referred to during my preparation). e.g. see here: https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20System%20Introspection.html

Reply:
This is an interesting point. I'm happy to leave it up to you but here are my thoughts:
- A couple of comments below I think it makes sense to keep the term "System Introspection" (ln 64), especially given that this is the term that CMake is using. This is okay in this context because it just means the system inspecting itself, which may be true depending on how broadly we are defining "system" and what exactly we mean by this. 
- Grammatically, it is incorrect to say "instrospecting something" as in "introspecting the system". Introspecting cannot take an object, because it always means something inspecting itself. (One can say that "X is introspecting" but not "X is introspecting Y".) Since we are talking about one thing (cmake) examining another ("the system" i.e. environment, OS, hardware etc. - in this context it is clearly something separate from cmake) then "inspecting" is the correct verb. 
- I would argue that the same is true for the comment below (ln 53). "Hardware introspection" doesn't make sense to me because the hardware does not inspect itself.

Reply:
@mmcleod89 Thank you. I have now used inspecting/inspection consistently throughout.